File: src/game/dayCycle.ts
// Synced to absolute UTC so every client sees the same cycle position.
// 10-minute period: 00:00 UTC start of cycle, 00:10 next, etc.
const CYCLE_MS = 10 * 60 * 1000;

// Returns 0..1 around the cycle. 0 = dawn, 0.25 = noon, 0.5 = dusk, 0.75 = midnight.
export function dayPhase(): number {
  return (Date.now() % CYCLE_MS) / CYCLE_MS;
}

// Smooth brightness curve — peaks at noon, troughs at midnight. Floor
// kept high so the world reads as a vibrant glowing biome at all hours,
// with the cycle adding a subtle pulse rather than a real day/night.
export function dayBrightness(phase: number): number {
  return 0.95 + 0.45 * (0.5 + 0.5 * Math.cos(2 * Math.PI * (phase - 0.25)));
}

// Hue rotation in radians. Smaller sweep (±π/12 ≈ ±15°) keeps the
// blue spawn region readable as blue throughout the cycle — only a
// gentle warm/cool drift, not a full hue swing into violet/magenta.
export function dayHueAngle(phase: number): number {
  return Math.cos(2 * Math.PI * (phase - 0.25)) * (Math.PI / 12);
}

File: LayToRest/Assets/Scripts/Core/DayCycle.cs
using System;
using UnityEngine;

namespace LayToRest.Core
{
    /// Port of src/game/dayCycle.ts — 10-minute cycle synced to absolute
    /// UTC so every client sees the same phase. 0 = dawn, 0.25 = noon,
    /// 0.5 = dusk, 0.75 = midnight.
    public class DayCycle : MonoBehaviour
    {
        const double CycleMs = 10.0 * 60.0 * 1000.0;

        [Tooltip("Optional sun — intensity is driven by the brightness curve.")]
        public Light sun;
        [Tooltip("Base intensity the brightness curve multiplies.")]
        public float sunBaseIntensity = 1f;

        public static float Phase
        {
            get
            {
                double ms = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() % CycleMs;
                return (float)(ms / CycleMs);
            }
        }

        /// Peaks at noon, troughs at midnight; floor kept high so the world
        /// always reads as a glowing biome.
        public static float Brightness(float phase) =>
            0.95f + 0.45f * (0.5f + 0.5f * Mathf.Cos(2f * Mathf.PI * (phase - 0.25f)));

        /// Gentle warm/cool hue drift, ±15°, in radians.
        public static float HueAngle(float phase) =>
            Mathf.Cos(2f * Mathf.PI * (phase - 0.25f)) * (Mathf.PI / 12f);

        void Update()
        {
            if (sun != null)
                sun.intensity = sunBaseIntensity * Brightness(Phase);
        }
    }
}
